README.md
# jsonbook

[Wiktionary](https://en.wiktionary.org/) dump in accessible `JSON` format.

## TL;DR

Just get the files: [Wiktionary dump in `JSON` format][filelink]

## Why

The English Wiktionary gracefully provides regular dumps of all its content
so anyone can easily parse the content for his/her own use. Unfortunately,
the _parsing_ part can be daunting and quickly turn any interested developer
off. This is because:

* The dump file is **very big** (5.5Gb uncompressed)
* The dump file contains many [other things](doc/dump_stats_en.md).
* Wiktionary is a _dictionary in English_, not a _English dictionary_, so
  its entries includes a lot of non-English words.
* The entries aren't in alphabetical order.  

In other words, to get to the content you want, you have first filter through
millions of pages you aren't remotely interested in. Or, you can spam the system
with lots of API calls.

## Solution

**Jsonbook** helps remove the first road block of your making
awesome use of the Wiktionary dump. It does the following:

1. Retrieve only the word articles
2. Organize all the articles by language
3. Convert the text to hierarchical tree
4. Save all the content to individual `JSON` files.

See a [sample output](doc/sample-output.tar.bz2) of the entry for "gratis".

Currently it takes about 58 minutes to parse the entire English Wiktionary
dump on a 2013 MacBook Pro.

[filelink]: https://drive.google.com/open?id=1X83TEseBImv8YXk7GoQOuGQt3qDMbiaI
[enwiktionary-link]: https://dumps.wikimedia.org/enwiktionary/latest/enwiktionary-latest-pages-articles.xml.bz2


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

# [0.2.0](https://github.com/dan1wang/jsonbook/compare/v0.1.0...v0.2.0) (2019-04-12)


### Bug Fixes

* use sytem temp dir & recursive mkDir ([e71b860](https://github.com/dan1wang/jsonbook/commit/e71b860))


### Features

* split xml to json ([e15f77c](https://github.com/dan1wang/jsonbook/commit/e15f77c))



# 0.1.0 (2019-04-11)


src/file.js
/* eslint-disable no-alert, no-console */
'use strict';

const fs = require('fs');
const path = require('path');
const sys = require('os');
const branch = require('git-branch');
const bgWhite = require('chalk').bgWhite;
const PassThrough = require('stream').PassThrough;

/**
 * Get a pass-through stream processor for displaying file loading status
 * @param {string} fPath File path
 * @param {string} fName File name
 * @return {object} A PassThrough object
 */
function getFileBar(fPath, fName) {
  let startTime = 0;
  const fTotalSize = fs.statSync(fPath + fName).size;
  let fReadSize = 0;
  let lastProgress = -1;
  const fileBar = new PassThrough();
  fileBar.on('data', (data) => {
    fReadSize += data.length;
    if (startTime ===0) startTime = (new Date()).valueOf();
    const progress = Math.round(1000 * fReadSize / fTotalSize) / 1000;
    if (progress !== lastProgress) {
      lastProgress = progress;
      // draw progress bar
      const filled = Math.round(progress * 20, 0);
      const bars = bgWhite(' '.repeat(filled)) + '.'.repeat(20 - filled);
      const pct = (progress * 100).toFixed(1) + '%';
      // process.stdout.clearLine();
      process.stdout.cursorTo(0);
      process.stdout.write(`[${bars}]: Reading "${fName}" (${pct})`);
    }
  });
  fileBar.on('end', () => {
    const elapsedTime = ((new Date()).valueOf() - startTime) / (60 * 1000);
    const h = Math.floor(elapsedTime / 60);
    const m = Math.floor(elapsedTime) % 60;
    let t;
    switch (h) {
      case 0: t = m + ' mins.'; break;
      case 1: t = h + ' hr ' + m + ' mins.'; break;
      default: t = h + ' hrs ' + m + ' mins.';
    }
    console.log('\nFinished in ' + t);
  });
  return fileBar;
}

/**
 * Get the file stream of a file
 * @param {string} fPath File path
 * @param {string} fName File name
 * @return {object} A ReadStream object
 */
function getFileStream(fPath, fName) {
  const fullFilePath = fPath + fName;
  if (fs.existsSync(fPath)) {
    const fStream = fs.createReadStream(fullFilePath);
    return fStream;
  } else {
    throw new Error(`Cannot find "${fullFilePath}".`);
  }
}

/**
 * Process an XML file.
 * If fName is empty, the first .xml file is processed
 * @param {string} fPath File path
 * @param {string} fName File name
 * @param {object} parser XML parser
 */
function parseXmlFile(fPath, fName, parser) {
  if (fPath[fPath.length-1] !== '/') fPath = fPath + '/';
  if (!fName) {
    const files = fs.readdirSync(fPath).sort();
    for (let i = 0; i < files.length; i++) {
      const f = files[i];
      if (f.slice(f.length-4).toLowerCase() === '.xml') {
        fName = f;
        break;
      }
    }
    if (!fName) {
      throw new Error(`Cannot find any XML file in "${fPath}".`);
    }
  }
  const fFileBar = getFileBar(fPath, fName);
  getFileStream(fPath, fName).pipe(fFileBar).pipe(parser);
}

/**
 * Sanitize filename
 * @param {string} fName Dirty file name
 * @return {string} Sanitized file name
 * Based on public-domain code by Sam Hocevar
 * https://github.com/parshap/node-sanitize-filename/blob/master/index.js
 */
function safeFileName(fName) {
  const illegalRe = /[\/\?<>\\:\*\|":]/g;
  const controlRe = /[\x00-\x1f\x80-\x9f]/g;
  const reservedRe = /^\.+$/;
  const windowsReservedRe = /^(con|prn|aux|nul|com[0-9]|lpt[0-9])(\..*)?$/i;
  const windowsTrailingRe = /[\. ]+$/;

  const sanitized = fName
      .replace(illegalRe, '-')
      .replace(controlRe, '')
      .replace(reservedRe, '')
      .replace(windowsReservedRe, '')
      .replace(windowsTrailingRe, '');
  return sanitized;
}

/**
 * Make sure a directory exists
 * @param {string} fPath target directory
 * @param {function} callback callback function
 */
function mkDir(fPath, callback) {
  fs.promises
      .mkdir(path.dirname(fPath), {recursive: true})
      .then( callback );
}

/**
 * Get temporary directory
 * @return {string} Directory path
 */
function getTempDir() {
  return sys.tmpdir() + '/jsonbook-' + branch.sync();
}

exports.safeFileName = safeFileName;
exports.parseXmlFile = parseXmlFile;
exports.mkDir = mkDir;
exports.getTempDir = getTempDir;


filecount.js
/* eslint-disable no-alert, no-console */
'use strict';

// code based on https://code-maven.com/list-content-of-directory-with-nodejs

const fs = require('fs');
const path = require('path');
const sys = require('os');

const fPath = sys.tmpdir() + '/jsonbook-master';

function getDirectories(source) {
  const dirs = fs.readdirSync(source)
      .map(name => path.join(source, name))
      .filter(source => fs.lstatSync(source).isDirectory())
  return dirs;
}

function countFiles(source) {
  const subdirs = getDirectories(source);
  let count = 0;
  if ((subdirs) && (subdirs.length)) {
    for (let i = 0; i < subdirs.length; i++) count+= countFiles(subdirs[i]);
  } else {
    count = fs.readdirSync(source).length;
  }
  return count;
}

const fPathLang = getDirectories(fPath);

for (let i = 0; i < fPathLang.length; i++) {
  fs.appendFileSync('filecount.tsv', fPathLang[i] + '\t' + countFiles(fPathLang[i]) + '\n');
}


src/parser.js
/* eslint-disable no-alert, no-console */
'use strict';

const sax = require('sax');

const struct = require('./wikiText/struct').struct;

const xParser = sax.createStream(
    true,
    {normalize: false, lowercase: false}
);

let article = {};
const xPath = [];
let xPathString = '';

xParser.on('text', (t) => {
  switch (xPathString) {
    case 'mediawiki/page/id': article['id'] = Number(t); break;
    case 'mediawiki/page/ns': article['ns'] = Number(t); break;
    case 'mediawiki/page/title': article['title'] = t; break;
    case 'mediawiki/page/revision/timestamp':
      article['timestamp'] = (new Date(t)).valueOf();
      break;
    case 'mediawiki/page/revision/text':
      article['text'] = struct(t);
      break;
    default:
  }
});

xParser.on('opentag', (node) => {
  if (xPath.length === 0) {
    if (node.name !== 'mediawiki') {
      throw new Error('XML file is not a Wiktionary file!');
    }
  }
  xPath.push(node.name);
  xPathString = xPath.join('/');
});

xParser.on('closetag', (node) => {
  if (xPathString === 'mediawiki/page') {
    if (article.ns === 0) {
      if (typeof xParser.processorFn === 'function') {
        xParser.processorFn(article);
      }
    }
    article = {};
  }
  xPath.pop(node.name);
  xPathString = xPath.join('/');
});

xParser.onerror = (e) => console.log('Error: ' + e);

exports.parser = xParser;


src/processor.en.js
/* eslint-disable no-alert, no-console, require-jsdoc */
'use strict';

const json = require('./json');
const fsUtils = require('./file');

const tmpDir = fsUtils.getTempDir();

let articleCount = 0;

exports.processor = (article) => {
  articleCount++;
  // const collection = (Math.floor(articleCount / 10000) + 1) * 10000;
  if (articleCount % 100000 === 0) {
    console.log(`checking in: ${articleCount} articles processed...`);
  }

  if ((article.text[0]) &&
      (article.text[0]['subSections']) &&
      (article.text[0]['subSections'].length)) {
    const topSection = article.text[0];
    const subSections = topSection.subSections;
    for (let i=0; i < subSections.length; i++) {
      const lang = subSections[i].title;
      const secArticle = Object.assign({}, article);
      secArticle.text = [{
        level: 0,
        title: '',
        text: topSection.text,
        subSections: [subSections[i]],
      }];
      const fPath = `${tmpDir}/${lang}/${article.title[0]}/${article.id}.json`;
      fsUtils.mkDir(fPath, () => json.save(fPath, article));
    }
  } else {
    const fPath = `${tmpDir}/__stubs/${article.id}.json`;
    fsUtils.mkDir(fPath, () => json.save(fPath, article));
  }
  // const title = fsUtils.safeFileName(article.title);
};


src/wikiText/struct.js
/* eslint-disable no-alert, no-console */
'use strict';

exports.struct = struct;

/**
 * Trim leading and ending line breaks
 *
 * "  \n  some text \n -----\n" should becomes "  some text"
 * @param {string} text
 * @return {string} a new string
 */
function trimText(text) {
  return text
      // trim leading line breaks (but not identation)
      .replace(/^\s*\n+/, '')
      // trim ending line breaks and horizontal rule
      .replace(/\s*\n[-]*\s*$/, '');
}

/**
 * Find all headings in an article text
 * @param {string} articleText
 * @return {array} headings infomation (with level, title, and position)
 */
function findHeadings(articleText) {
  const regexHeading = new RegExp('^\\s*(=+)(.*)\\1\\s*$', 'gm');
  const results = [];
  let match;
  while ((match = regexHeading.exec(articleText)) !== null) {
    results.push({
      line: match[0],
      level: match[1].length,
      title: match[2].trim(),
      pos: regexHeading.lastIndex - match[0].length,
    });
  }
  return results;
}

/**
 * Return a new array with a level 0 section, followed by
 * all headings with section text
 * @param {array} headings heading information
 * @param {string} articleText original article text
 * @return {array} all headings (with level, title, and position)
 */
function buildTextList(headings, articleText) {
  if ((headings) && (headings.length)) {
    const top = {
      level: 0,
      title: '',
      text: trimText(articleText.substring(0, headings[0].pos)),
    };
    const results = [top];
    const lastIndex = headings.length-1;
    for (let i=0; i < headings.length; i++) {
      const currH = headings[i];
      const secStart = currH.pos + currH.line.length;
      const secEnd = (i !== lastIndex) ? headings[i+1].pos: articleText.length;
      const secText = trimText(articleText.substring(secStart, secEnd));
      results.push({level: currH.level, title: currH.title, text: secText});
    }
    return results;
  } else {
    return [{level: 0, title: '', text: trimText(articleText)}];
  }
}

/**
 * Turn a flat array of text to hierarchical array
 * @param {array} flatText A flat array of text
 * @return {array}
 */
function buildTextTree(flatText) {
  if ((!flatText instanceof Array)||(flatText.length === 0 )) return null;
  const results = [Object.assign({}, flatText[0])];
  let topIndex = 0;
  let topLevel = flatText[0].level;
  let childCount = 0;
  for (let i=1; i < flatText.length; i++) {
    if (flatText[i].level > topLevel) {
      childCount++;
    } else {
      if (childCount > 0) {
        const children = flatText.slice(topIndex+1, topIndex+childCount+1);
        results[results.length-1]['subSections'] = buildTextTree(children);
      }
      childCount = 0;
      topIndex = i;
      topLevel = flatText[i].level;
      results.push(Object.assign({}, flatText[i]));
    }
  }
  if (childCount > 0) {
    const children = flatText.slice(topIndex+1, topIndex+childCount+1);
    results[results.length-1]['subSections'] = buildTextTree(children);
  }
  return results;
}


/**
 * Turn flat article text to hierarchical text
 * @param {string} aText Original article text
 * @return {array}
 */
function struct(aText) {
  return buildTextTree(buildTextList(findHeadings(aText), aText));
}


index.js
/* eslint-disable no-alert, no-console */
'use strict';

const fs = require('fs');
const sax = require('sax');
const chalk = require("chalk");

const DATA_DIR = 'data/';

function getFileStream() {
  const files = fs.readdirSync(DATA_DIR);
  for (let i = 0; i < files.length; i++) {
    const fName = files[i];
    if (fName.slice(fName.length-4).toLowerCase() === '.xml') {
      const reader = fs.createReadStream(DATA_DIR + fName);
      const startTime = (new Date()).valueOf();
      let fTotalSize = fs.statSync(DATA_DIR + '/' + fName).size;
      let fReadSize = 0;
      let lastProgress = -1;
      reader.on('data', (data) => {
        fReadSize += data.length;
        const progress = Math.round(1000 * fReadSize / fTotalSize) / 1000;
        if (progress > lastProgress) {
          lastProgress = progress;
          drawProgressBar(progress, `Reading '${fName}'...`);
        }
      });
      reader.on('end', () => {
        const endTime = (new Date()).valueOf();
        const elapsed = (endTime - startTime) / (60 * 1000);
        const h = Math.floor(elapsed / 60);
        const m = Math.floor(elapsed) % 60;
        console.log(`\nFinished in ${h} h ${m} min.`);
      });
      return reader;
    }
  }
  throw new Error('Cannot find any .xml file in "data/" directory.');
}

function drawProgressBar(progress, message) {
  const bars = Math.round(progress * 20,0);
  const filled = chalk.bgWhite(' '.repeat(bars));
  const empty = '.'.repeat(20 - bars);
  const pct = (Math.floor(progress * 1000)/10) + '%';
  process.stdout.clearLine();
  process.stdout.cursorTo(0);
  process.stdout.write(`[${filled}${empty}]: ${message} (${pct})`);
}

const xParser = sax.createStream(
    true,
    {normalize: false, lowercase: false}
);

const xPath = [];
let xPathString = '';
let article = {};

xParser.on('text', (t) => {
  switch (xPathString) {
    case 'mediawiki/page/id': article['id'] = Number(t); break;
    case 'mediawiki/page/ns': article['ns'] = Number(t); break;
    case 'mediawiki/page/title': article['title'] = t; break;
    case 'mediawiki/page/revision/text': article['text'] = t; break;
    case 'mediawiki/page/revision/timestamp':
      article['timestamp'] = (new Date(t)).valueOf();
      break;
    default:
  }
});

xParser.on('opentag', (node) => {
  if (xPath.length === 0) {
    if (node.name !== 'mediawiki') {
      throw new Error('XML file is not a Wiktionary file!');
    }
  }
  xPath.push(node.name);
  xPathString = xPath.join('/');
});

xParser.on('closetag', (node) => {
  if (xPathString === 'mediawiki/page') {
    logNonDictionaryEntries(article);
    if (article.ns === 0) {
      logTag(article);
      logTemplate(article);
      logMagic(article);
      logEnUncategorized(article);
    }
    article = {};
  }
  xPath.pop(node.name);
  xPathString = xPath.join('/');
});

const logIt = fs.appendFileSync;

function hasString(str, subStr) {
  return (str.indexOf(subStr, 0) !== -1)
}

function logNonDictionaryEntries(article) {
  if (article.ns !== 0) {
    let logFile;
    switch (article.ns) {
      case   4: logFile = '004 Wiktionary.tsv'; break;
      case   6: logFile = '006 file.tsv'; break;
      case   8: logFile = '008 MediaWiki.tsv'
      case  10: logFile = '010 template.tsv'; break;
      case  12: logFile = '012 help.tsv'; break;
      case  14: logFile = '014 category.tsv'; break;
      case  90: logFile = '090 thread.tsv'; break;
      case  92: logFile = '092 summary.tsv'; break;
      case 100: logFile = '100 appendix.tsv'; break;
      case 102: logFile = '102 condordance.tsv'; break;
      case 104: logFile = '104 index.tsv'; break;
      case 106: logFile = '106 rhymes.tsv'; break;
      case 108: logFile = '108 transwiki.tsv'; break;
      case 110: logFile = '110 thesaurus.tsv'; break;
      case 114: logFile = '114 citation.tsv'; break;
      case 116: logFile = '116 sign gloss.tsv'; break;
      case 118: logFile = '118 reconstruction.tsv'; break;
      case 828: logFile = '828 module.tsv'; break;
      default: logFile = article.ns + '.tsv';
    }
    const line = article.title + '\t' + article.id + '\n'
    logIt(logFile, line);
  }
}

function logEnUncategorized(article) {
  const str = article.text;
  if (
    (hasString(str,'==English==')) &&
    (!hasString(str,'{{en-')) &&
    (!hasString(str,'{{catlangname|en')) &&
    (!hasString(str,'{{cln|en')) &&
    (!hasString(str,'{{head|en')) &&
    (!hasString(str,'[[Category:English'))
  ) {
    const logFile = 'en - uncategorized.tsv';
    const line = article.title + '\t' + article.id + '\n'
    logIt(logFile, line);
  }
}

function logTag(article) {
  const str = article.text;
  const line = article.title + '\t' + article.id + '\n'
  if (hasString(str,'<!--')) logIt('all - tag - comment.tsv', line);
  if (hasString(str,'<pre')) logIt('all - tag - pre.tsv', line);
  if (hasString(str,'<code>')) logIt('all - tag - code.tsv', line);
}

function logTemplate(article) {
  const str = article.text;
  const line = article.title + '\t' + article.id + '\n'
  if (hasString(str,'{{}}')) logIt('all - temp - empty.tsv', line);
  if (hasString(str,'{{\\}}')) logIt('all - temp - empty 2.tsv', line);
}

function logMagic(article) {
  const str = article.text;
  const found = str.match(/__[A-Z]+__/g);
  if (found) {
    const line = article.title + '\t' + article.id + '\t' + found.join('\t') + '\n';
    logIt('all - magic - switch.tsv', line);
  }
}



// -------------------------------------------


getFileStream().pipe(xParser);
